src/bindings.cpp

```diff
@@ -21,51 +22,97 @@ extern "C"
     std::unique_ptr<Player> player = nullptr;
     std::unique_ptr<Analyzer> analyzer = std::make_unique<Analyzer>(2048);
 
-    /// @brief Set a dart function to call when the sound with [handle] handle ends
-    /// @param callback this is the dart function that will be called
-    ///     when the sound ends to play.
-    ///     Must be global or a static class member:
-    ///     ```@pragma('vm:entry-point')
-    ///        void playEndedCallback(int handle) {
-    ///             // here the sound with [handle] has ended.
-    ///             // you can play again
-    ///             soLoudController.soLoudFFI.play(handle);
-    ///             // or dispose it
-    ///             soLoudController.soLoudFFI.stop(handle);
-    ///        }
-    ///     ```
-    /// @param handle the handle to the sound
-    /// @param callback this is the dart function that will be called
-    ///         when the sound ends to play
-    /// @return true if success;
-    // FFI_PLUGIN_EXPORT bool setPlayEndedCallback(void (*callback)(unsigned int), unsigned int handle)
-    // {
-    //     if (!player.get()->isInited()) return false;
-    //     ActiveSound* sound = player.get()->findByHandle(handle);
-    //     if (sound != nullptr) {
-    //         sound->playEndedCallback = callback;
-    //         return true;
-    //     }
-    //     return false;
-    // }
+    typedef void (*dartVoiceEndedCallback_t)(unsigned int *);
+    typedef void (*dartFileLoadedCallback_t)(enum PlayerErrors *, char *completeFileName, unsigned int *);
+    typedef void (*dartStateChangedCallback_t)(enum PlayerStateEvents *);
+
+    // to be used by `NativeCallable`, these functions must return void.
+    void (*dartVoiceEndedCallback)(unsigned int *) = nullptr;
+    void (*dartFileLoadedCallback)(enum PlayerErrors *, char *completeFileName, unsigned int *) = nullptr;
+    void (*dartStateChangedCallback)(enum PlayerStateEvents *) = nullptr;
+
+    FFI_PLUGIN_EXPORT void nativeFree(void *pointer) {
+        free(pointer);
+    }
+    
+    /// The callback to monitor when a voice ends.
+    ///
+    /// It is called by void `Soloud::stopVoice_internal(unsigned int aVoice)` when a voice ends.
+    void voiceEndedCallback(unsigned int *handle)
+    {
+        if (dartVoiceEndedCallback == nullptr)
+            return;
+        player->removeHandle(*handle);
```

Comment:
What does it mean when `player` exists in C++ land and `dartVoiceEndedCallback` is null? Should we not remove the handle with `player->removeHandle()` in that case? This looks like we're sweeping a problem under the rug and it could bite us later.

Reply:
The `dartVoiceEndedCallback` is initialized in the below `initEngine()` and set to null in the `deinit()`. If `dartVoiceEndedCallback` is not null, also the player is not null. This means that if `dartVoiceEndedCallback` is null the player does not exist.

Reply:
Fair enough. Thanks for the explanation!